core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2333,3 +2333,34 @@ private[spark] class RedirectThread(
     }
   }
 }
+
+/**
+ * Circular buffer, which consume all of the data write to it.
```

Comment:
An [[OutputStream]] that will store the last 10 kilobytes written to it in a circular buffer.  The current contents of the buffer can be accessed using the `toString` method.